### ISSUE TITLE
How to change pressure level in zonal mean 2d plots?

### ISSUE THREAD

**Issue by kaabre** [E3SM-Project/e3sm_diags#182]
ie, plots where the pressure level is not constant.  I would like to start at smaller pressure levels than 100 mb which is currently the default (and seems to be hard-coded).  Is there a way to do this without having to resort to using a development environment and changing the source code?  The regular 'plevs' parameter in the cfg file does not do this (it creates several different plots with constant pressure level, and in the zonal mean 2d plots it actually breaks the output)

I'm looking in the driver file: https://github.com/E3SM-Project/acme_diags/blob/5b192f779e0ce0953459ed1a23762e51f3653c3d/acme_diags/driver/zonal_mean_2d_driver.py

Thanks!

**Comment by golaz** [comment]
:+1: We should have the flexibility to do this. I didn't realize it doesn't work.

**Comment by chengzhuzhang** [comment]
Hi Lee @kaabre , Thank you for the suggestions. I will add a new parameter for user to specify pressure levels for zonal mean 2d plots. Our team also talked about adding an option to make the pressure axis linear or log. I will have this done together with this issue. Some aesthetic enhancement is also needed for this plot set..I will let you know how it goes. Thanks!

**Comment by zshaheen** [comment]
Implementing this isn't hard to do at all, but it'll cause an issue with the provenance.

Recall that for every plot, we have the provenance, which is a command to recreate the image.

![screen shot 2018-12-13 at 12 25 41 pm](https://user-images.githubusercontent.com/4239938/49965470-1ee96680-fed2-11e8-8e66-24c41a7f1025.png)

With a parameter that uses `numpy`, parsing it via the command line is hard to do. A possible workaround is to just list all of the parameters via a Python file. A user will then just copy them into `something.py` and to recreate the image, just run `e3sm_diags -p something.py` instead of the very long command.

**Comment by zshaheen** [comment]
@kaabre My above post might be confusing. Don't worry if it doesn't make sense to you. We have things mostly working, but need to account for the above point.

**Comment by oksanaguba** [comment]
Would you please give me an example how to use different pressure levels (deeper top)? i have been trying variations of  this string
```
zonal_mean_2d_plevs = [30.,50.,70.,100.,150.,200.,250.,300.,400.,500.,600.,700.,775.,850.,925.,1000.]
```
but output is always at 100 mb and stdout is always
```
Selected pressure level: [ 100.          115.47819847  133.35214322  153.99265261  177.827941
  205.35250265  237.13737057  273.84196343  316.22776602  365.17412725
  421.69650343  486.96752517  562.34132519  649.38163158  749.89420933
  865.96432336 1000.        ]
```

**Comment by chengzhuzhang** [comment]
Hi, @oksanaguba, the zonal_mean_2d_plev parameter was implemented on a branch zonal_mean_2d_plev_iss182, we didn't merge the feature because we wanted to make more changes (enabling users to use numpy array for plevs) but didn't get to it yet. Let me create a PR, at least the user can use this new parameter.

**Comment by oksanaguba** [comment]
Thanks! This would be great!

**Comment by zshaheen** [comment]
This was merged into master on 0fe04a8aa6b979c8b207223313b55bb731c35d9e. Below is a sample output. Since the y-axis is a little cluttered, you might want to [modify the corresponding script that does the plotting for that plotset (zonal_mean_2d_plot.py)](https://github.com/E3SM-Project/e3sm_diags/blob/master/acme_diags/plot/cartopy/zonal_mean_2d_plot.py).

You'll need to install the latest code from master. [Please see the documentation here to create an environment for this.](https://e3sm-project.github.io/e3sm_diags/docs/html/install.html#environment-for-development)

![-T-ANN-global-1](https://user-images.githubusercontent.com/4239938/58820768-51099480-85e8-11e9-9b07-e9e7ecd7fcbf.png)

---

The reason we didn't immediately do this is because we're (still) thinking of a way to reconcile the various plot set specific parameters. Obviously, the parameter `zonal_mean_2d_plevs ` is only for the `'zonal_mean_2d'` plot set. However, with the current implementation, this is a "valid" parameter for all of the plot sets.

When `e3sm_diags` becomes the only diagnostics package in the climate community with thousands of plot sets, you can imagine that we'd need some *plot-set-specific* parameters. In other words, instead of prefixing each parameter with `{plot_set_name}_parameter`, there should be a better, more organized way to do this.

**Comment by oksanaguba** [comment]
Thanks, @zshaheen -- I am using anvil, where (I assume) the repo is already cloned by your team. If so, do I have to follow installation steps (which I did not follow before on anvil) or could anvil repo be updated instead by you? At the moment plevs feature does not work on anvil for me (I tried today). Thanks!

**Comment by zshaheen** [comment]
@oksanaguba Are you using the unified env on Anvil? If so, then sadly you'd need to wait for it to be updated by Xylar and us (when we release a version of the software with the stuff on master. We can do this immediately if needed).

However, if you do not want to wait on anyone, I'd just follow the instructions that I linked, [creating an environment for development](https://e3sm-project.github.io/e3sm_diags/docs/html/install.html#environment-for-development)) It should only take 10 mins or so. When you have that done, you'll have the latest code on master installed and you can run it with the `zonal_mean_2d_plevs` parameter.

**Comment by xylar** [comment]
@oksanaguba, just a little more background, the process of updating E3SM-Unified is not trivial so it only happens every few months when several of our shared packages have been updated.  This means you don't typically get the latest development from any of them unless you install them as @zshaheen suggests.

**Comment by oksanaguba** [comment]
@xylar @zshaheen I tied to follow steps as in instructions, loaded anaconda/4.4.0 on blues node (not b512) and got error below at that particular step. What should I do? Thanks!
```
[onguba@blueslogin4 ~]$ conda env create -f e3sm_diags_env.yml
Using Anaconda API: https://api.anaconda.org
Fetching package metadata .................
Solving package specifications: .

UnsatisfiableError: The following specifications were found to be in conflict:
  - cartopy 0.17.0*
  - curl 7.64.1*
  - gstreamer 1.14.4*
  - sip 4.19.8*
  - sqlite 3.26.0*
Use "conda info <package>" to see the dependencies for each package.
```

**Comment by oksanaguba** [comment]
oh and I got at some point earlier the warning below, does it matter?
```
[WARNING] "ssl_verify" is not a known configuration key
```

**Comment by oksanaguba** [comment]
please, ignore my prev. messages for now, I skipped condo update condo step, repeating now...

**Comment by xylar** [comment]
@oksanaguba, the first step, before any of this would be to install Miniconda3 in your home directory on blues if you haven't already done that.  Please google that and follow the instructions.  Then, you can install your own environments from Miniconda3.  With the system python, you may not be able to create your own environments or update conda or anything like that.

**Comment by oksanaguba** [comment]
Thanks @xylar . I followed this website, installed miniconda3 https://conda.io/projects/conda/en/latest/user-guide/install/index.html , did not load system's anaconda, tried to follow instructions and still got 
```
(base) [onguba@blueslogin2 ~]$ conda env create -f e3sm_diags_env.yml
Collecting package metadata: done
Solving environment: failed

UnsatisfiableError: The following specifications were found to be in conflict:
  - e3sm_diags=1.7.0
  - openssl=1.1.1b
  - pcre=8.41
  - sip=4.19.8
  - sqlite=3.26.0
Use "conda search <package> --info" to see the dependencies for each package.
```
what am I doing wrong?

**Comment by oksanaguba** [comment]
I closed the session and logged in again and then did not run 'clean' step, got fewer conflicts:
```
(base) [onguba@blueslogin1 ~]$ conda env create -f e3sm_diags_env.yml
Collecting package metadata: done
Solving environment: failed

UnsatisfiableError: The following specifications were found to be in conflict:
  - glib=2.58.3
  - pyqt=5.9.2
  - sqlite=3.26.0
Use "conda search <package> --info" to see the dependencies for each package.
```

**Comment by oksanaguba** [comment]
... and more. I would greatly appreciate some help with this. As you can see I have no experience with python/conda.
```
(base) [onguba@blueslogin1 ~]$ conda search glib
Loading channels: done
# Name                       Version           Build  Channel             
glib                          2.43.0               0  pkgs/free           
glib                          2.43.0               1  pkgs/free           
glib                          2.50.2               1  pkgs/free           
glib                          2.53.6      h0d2cc06_1  pkgs/main           
glib                          2.53.6      h56f7bfe_1  pkgs/main           
glib                          2.53.6      h5d9569c_2  pkgs/main           
glib                          2.53.6      h7684f33_1  pkgs/main           
glib                          2.53.6      hc861d11_1  pkgs/main           
glib                          2.56.0      h000015b_0  pkgs/main           
glib                          2.56.1      h000015b_0  pkgs/main           
glib                          2.56.2      hd408876_0  pkgs/main           
(base) [onguba@blueslogin1 ~]$ conda search glib=2.58.3
Loading channels: done
No match found for: glib=2.58.3. Search: *glib*=2.58.3

PackagesNotFoundError: The following packages are not available from current channels:

  - glib=2.58.3

Current channels:

  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/free/linux-64
  - https://repo.anaconda.com/pkgs/free/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.


(base) [onguba@blueslogin1 ~]$ conda search *glib*=2.58.3
Loading channels: done

PackagesNotFoundError: The following packages are not available from current channels:

  - *glib*=2.58.3

Current channels:

  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/free/linux-64
  - https://repo.anaconda.com/pkgs/free/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.
```

**Comment by xylar** [comment]
@oksanaguba, I would suggest opening a new issue on `e3sm_diags` for this since this issue is unrelated and has been closed already.

That being said, I can reproduce the error you report as well so I think @zshaheen or @chengzhuzhang will need to update the yaml file to fix the conflicts.

**Comment by xylar** [comment]
Please continue the conversation on #224.
